File: legacy-python/app/totp.py
"""TOTP (RFC 6238) cho 2FA — thuần stdlib, tương thích Google Authenticator.

Chu kỳ 30s, 6 số, HMAC-SHA1 (chuẩn mặc định của Google Authenticator).
"""
import base64
import hashlib
import hmac
import secrets
import struct
import time
import urllib.parse


def new_secret() -> str:
    """Secret base32 20 byte (chuẩn GA)."""
    return base64.b32encode(secrets.token_bytes(20)).decode().rstrip("=")


def _code_at(secret: str, counter: int) -> str:
    # padding lại base32 (secret lưu không có '=')
    pad = "=" * (-len(secret) % 8)
    key = base64.b32decode(secret.upper() + pad)
    msg = struct.pack(">Q", counter)
    digest = hmac.new(key, msg, hashlib.sha1).digest()
    offset = digest[-1] & 0x0F
    num = struct.unpack(">I", digest[offset:offset + 4])[0] & 0x7FFFFFFF
    return f"{num % 1_000_000:06d}"


def totp_now(secret: str, at: float = None) -> str:
    t = int((at if at is not None else time.time()) // 30)
    return _code_at(secret, t)


def verify_totp(secret: str, code: str, window: int = 1) -> bool:
    """So khớp code với cửa sổ ±window chu kỳ (lệch đồng hồ điện thoại)."""
    code = (code or "").strip().replace(" ", "")
    if not code.isdigit() or len(code) != 6 or not secret:
        return False
    t = int(time.time() // 30)
    return any(hmac.compare_digest(_code_at(secret, t + off), code)
               for off in range(-window, window + 1))


def otpauth_uri(secret: str, email: str, brand_name: str = "ANS Music") -> str:
    """URI otpauth:// — Google Authenticator hiển thị '{brand} - {email}'."""
    label = urllib.parse.quote(f"{brand_name} - {email}")
    issuer = urllib.parse.quote(brand_name)
    return f"otpauth://totp/{label}?secret={secret}&issuer={issuer}&algorithm=SHA1&digits=6&period=30"

File: backend/src/totp.ts
/** TOTP (RFC 6238) cho 2FA — port từ backend/app/totp.py, dùng thư viện otplib.
 *
 *  Chu kỳ 30s, 6 số, HMAC-SHA1 (chuẩn mặc định Google Authenticator). otplib với
 *  các mặc định (algorithm=SHA1, digits=6, step=30, keyEncoder=base32) sinh/kiểm
 *  code Y HỆT thuật toán stdlib của Python → tương thích secret cũ trong DB. */
import { authenticator } from 'otplib';

/** Secret base32 20 byte (khớp Python: base32(token_bytes(20)) → 32 ký tự). */
export function newSecret(): string {
  return authenticator.generateSecret(20);
}

/** Code hiện tại cho secret (chu kỳ 30s hiện thời). */
export function totpNow(secret: string): string {
  return authenticator.generate(secret);
}

/** So khớp code với cửa sổ ±window chu kỳ (bù lệch đồng hồ điện thoại). */
export function verifyTotp(secret: string, code: string, window = 1): boolean {
  const clean = (code || '').trim().replace(/ /g, '');
  if (!/^\d{6}$/.test(clean) || !secret) return false;
  authenticator.options = { window };
  try {
    return authenticator.verify({ token: clean, secret });
  } catch {
    return false;
  }
}

/** URI otpauth:// — Google Authenticator hiển thị nhãn '{brand} - {email}'. */
export function otpauthUri(secret: string, email: string, brandName = 'ANS Music'): string {
  const label = encodeURIComponent(`${brandName} - ${email}`);
  const issuer = encodeURIComponent(brandName);
  return `otpauth://totp/${label}?secret=${secret}&issuer=${issuer}&algorithm=SHA1&digits=6&period=30`;
}
